### PR TITLE
check the correct size of per cpu maps in bpf syscall

### PR DESCRIPTION
**Pull request by mazdakn**
## Release Note

```release-note
None required
```

### CODE REVIEW

File: felix/bpf/bpf_syscall.go [hunk 1]
```diff
@@ -269,8 +269,16 @@ func checkMapIfDebug(mapFD MapFD, keySize, valueSize int) error {
 		if keySize != mapInfo.KeySize {
 			log.WithField("mapInfo", mapInfo).WithField("keyLen", keySize).Panic("Incorrect key length")
 		}
-		if valueSize >= 0 && valueSize != mapInfo.ValueSize {
-			log.WithField("mapInfo", mapInfo).WithField("valueLen", valueSize).Panic("Incorrect value length")
+		// The actual size of per cpu maps is equal to the value size * number of cpu
+		if mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_HASH || mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_ARRAY ||
+			mapInfo.Type == unix.BPF_MAP_TYPE_LRU_PERCPU_HASH || mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_CGROUP_STORAGE {
+			if valueSize >= 0 && valueSize != mapInfo.ValueSize*runtime.NumCPU() {
+				log.WithField("mapInfo", mapInfo).WithField("valueLen", valueSize).Panic("Incorrect value length")
```
Comment (tomastigera) [inline felix/bpf/bpf_syscall.go:276]: just to give a bit more info

```suggestion
				log.WithField("mapInfo", mapInfo).WithField("valueLen", valueSize).Panic("Incorrect value length for per-CPU map")
```

File: felix/bpf/bpf_syscall.go [hunk 2]
```diff
@@ -269,8 +269,16 @@ func checkMapIfDebug(mapFD MapFD, keySize, valueSize int) error {
 		if keySize != mapInfo.KeySize {
 			log.WithField("mapInfo", mapInfo).WithField("keyLen", keySize).Panic("Incorrect key length")
 		}
-		if valueSize >= 0 && valueSize != mapInfo.ValueSize {
-			log.WithField("mapInfo", mapInfo).WithField("valueLen", valueSize).Panic("Incorrect value length")
+		// The actual size of per cpu maps is equal to the value size * number of cpu
+		if mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_HASH || mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_ARRAY ||
```
Comment (tomastigera) [inline felix/bpf/bpf_syscall.go:273]: A suggestion, smaller, easier to add values in the future.

```suggestion
		switch (mapInfo.Type) {
		case unix.BPF_MAP_TYPE_PERCPU_HASH, unix.BPF_MAP_TYPE_PERCPU_ARRAY:
		    ...
		default:
		    ...
```

**Review by tomastigera — APPROVED** [review]
some nits, but LGTM

File: felix/bpf/bpf_syscall.go [hunk 3]
```diff
@@ -269,8 +269,16 @@ func checkMapIfDebug(mapFD MapFD, keySize, valueSize int) error {
...
+			mapInfo.Type == unix.BPF_MAP_TYPE_LRU_PERCPU_HASH || mapInfo.Type == unix.BPF_MAP_TYPE_PERCPU_CGROUP_STORAGE {
+			if valueSize >= 0 && valueSize != mapInfo.ValueSize*runtime.NumCPU() {
```
Comment (fasaxc) [inline felix/bpf/bpf_syscall.go:275]: Is `NumCPU()` the right number?  ISTR that the map gets sized according to some kernel-internal view of hte max number of CPUs that your motherboard _could_ support.

Comment (mazdakn) [reply felix/bpf/bpf_syscall.go:275]: hmmm... so if a system has two cpu sockets but only one is used, then the per cpu map should be provisioned for both? It not easy to find such a system for test.

Comment (tomastigera) [reply felix/bpf/bpf_syscall.go:275]: As @fasaxc suggested We should either use the libbpf function or get the value from /sys/devices/system/cpu/possible ourselves

Comment (mazdakn) [reply felix/bpf/bpf_syscall.go:275]: IIUC, felix does not have access to read `/sys/devices/system/cpu/possible` (is that correct?). If that's correct, does it mean the only option is by using libbpf?

Comment (tomastigera) [reply felix/bpf/bpf_syscall.go:275]: That library also read that file.

Comment (fasaxc) [reply felix/bpf/bpf_syscall.go:275]: Why do you think we don't have access to that file?

Comment (mazdakn) [reply felix/bpf/bpf_syscall.go:275]: Right, I thought of something else. Will implement by reading `/sys/devices/system/cpu/possible`

Comment (fasaxc) [reply felix/bpf/bpf_syscall.go:275]: Sure you don't want to just expose the libbpf function?  That should be easier given all the libbpf wrappers we already have.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**